src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs

```diff
@@ -276,17 +280,50 @@ public LoadState(IReadOnlyDictionary<string, ProjectId> projectPathToProjectIdMa
             public void Add(ProjectInfo info)
             {
                 _projectIdToProjectInfoMap.Add(info.Id, info);
-                _projectInfoList.Add(info);
+                //Memoized results of ProjectAlreadyReferencesProject may no longer be correct;
+                //reset the cache.
+                ProjectAlreadyReferencesProjectMemo = new Dictionary<ProjectId, Dictionary<ProjectId, bool>>();
```

Comment:
This is a reset of the cache of memoized results (see the comment above), so I don't think allocations in the constructor/initializer will be effective.